src/state.cpp

```diff
@@ -148,6 +148,44 @@ void pop_all()
     TRACE_FUNC_END;
 }
 
+bool check_states(const states::StateId state_given, const states::StateId state_to_check)
```

Comment:
I don't think there's any need for a helper function or recursion.

Why not just do a reverse iteration over the states, and pop each one until we are at the one we want to pop until?